mediaapi/routing/preview_url.go

```diff
@@ -0,0 +1,400 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package routing
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/hex"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	"golang.org/x/net/html"
+
+	"github.com/matrix-org/dendrite/clientapi/jsonerror"
+	"github.com/matrix-org/dendrite/mediaapi/fileutils"
+	"github.com/matrix-org/dendrite/mediaapi/storage"
+	"github.com/matrix-org/dendrite/mediaapi/types"
+	"github.com/matrix-org/dendrite/setup/config"
+	"github.com/matrix-org/util"
+	log "github.com/sirupsen/logrus"
+)
+
+// the parameters included in the incoming preview_url request
+type PreviewUrlRequest struct {
+	url url.URL
+	ts  types.UnixMs
+}
+
+// metadata of the url (media/html) to be previewed
+type mediaInfo struct {
+	MediaMetadata *types.MediaMetadata
+	Logger        *log.Entry
+}
+
+// PreviewUrlResponse defines the format of the JSON response
+// https://matrix.org/docs/spec/client_server/latest#get-matrix-media-r0-preview-url
+// TODO: add more fields to the response
+type PreviewUrlResponse struct {
+	MatrixImageSize int64  `json:"matrix:image:size"`
+	OgImage         string `json:"og:image"`
+	OgSiteName      string `json:"og:site_name"`
+	OgType          string `json:"og:type"`
+	OgTitle         string `json:"og:title"`
+	OgUrl           string `json:"og:url"`
+	OgDescription   string `json:"og:description"`
+}
+
+// PreviewUrl implements GET /preview_url.
+// Current implementation gets the url, parses the meta tags to obtain OG
+// data and returns a JSONResponse for the client to process.
+// TODO: avoid this endpoint in encrypted rooms
+func PreviewUrl(
+	req *http.Request, cfg *config.MediaAPI, db storage.Database) util.JSONResponse {
+
+	// parsing the request
+	r, err := parseRequest(req)
+	if err != nil {
+		return *err
+	}
+
+	// downloading metadata for the url
+	mediaInfo, err := r.downloadUrl(cfg, req.Context(), db)
+	if err != nil {
+		return *err
+	}
+
+	// preparing the response
+	response, err := mediaInfo.prepareResponse(cfg)
+	if err != nil {
+		return *err
+	}
+
+	return util.JSONResponse{
+		Code: http.StatusOK,
+		JSON: response,
+	}
+}
+
+// parseRequest parses the incoming preview request to extract the url and ts.
+// Returns either a parsed PreviewUrlRequest or error formatted as
+// util.JSONResponse
+func parseRequest(req *http.Request) (*PreviewUrlRequest, *util.JSONResponse) {
+
+	// get the url to be previewed from the request
+	urlToPreview := req.URL.Query().Get("url")
+	if len(urlToPreview) == 0 {
+		return nil, &util.JSONResponse{
+			Code: http.StatusBadRequest,
+			JSON: jsonerror.MissingArgument("Missing URL"),
+		}
+	}
+
+	// parse the url
+	parsedUrl, err := url.Parse(urlToPreview)
```

Comment:
We need to check this URL against a blacklist of URLs otherwise this is just an arbitrary proxy which is 😨 

See https://github.com/matrix-org/synapse/blob/4b965c862dc66c0da5d3240add70e9b5f0aa720b/synapse/rest/media/v1/preview_url_resource.py#L202 and https://github.com/matrix-org/synapse/blob/240b3ce253b2b36d79a0a6cc18aefd70ad066de3/synapse/config/repository.py#L323